pallets/providers/README.md

```diff
@@ -378,3 +377,45 @@ Error thrown when trying to get a root from a Main Storage Provider without pass
+
+The grace period is based on the total stake/capacity of the Storage Provider. In essence, the more stake a Storage Provider has, the longer the grace period.
+This is to avoid a high stake Storage Provider from being removed from the network prematurely.
+
+The runtime will automatically process any expired grace periods within the `on_poll` hook to ensure that the redundancy process is initiated as soon as possible. For every insolvent Storage Provider, an event will be emitted to notify the network and also mark the Storage Provider as insolvent, rendering them unable to operate as a Storage Provider. Finally all the of the Storage Provider's stake will be slashed and transfered to the treasury.
+
+### Ensuring Data Redundancy
+
+> [!IMPORTANT]
+> The runtime cannot ensure that all the data stored from an insolvent storage provider would be recovered. It is up to users and storage providers to ensure data redundancy since the runtime has no knowledge of file keys stored by whom.
+
+In the event when a BSP would become insolvent, the entire network of BSPs are responsible to regain data redundancy for the data they lost.
+
+To accomplish this, an off-chain indexer is required to discover the file keys which were stored by that insolvent Storage Provider. This is necessary since the runtime does not hold any file key in data in storage.
+
+Any account can call the `add_redundancy` extrinsic which requires a proof of inclusion of a given file key and the number of required BSPs needed to fulfill this request. The root is checked to be a current Bucket or BSP’s forest root to ensure that the file key does indeed exist as part of a Storage Provider's forest.
+
+This creates a traditional storage request with the specified amount of BSPs required. The caller of the extrinsic can optionally pass a list of data servers for the file key, which then will be marked in the storage request for the volunteers to request the data from. The caller is be able to obtain this information from the off-chain indexer.
+
+If the file was originally stored by an MSP, it is up to the user of the lost file or files within a bucket to execute the `transfer_file` or `transfer_bucket` extrinsics exposed by the file system pallet to move the data to a new MSP.
+
+#### Incentives and Storage Cleanup
+
+For every file key submitted for redundancy which was stored by an insolvent Storage Provider, the caller will be rewarded with a configurable amount of tokens which must be less than the slash factor to prevent abuse. The runtime will accrue the file size of each file key submitted for redundancy for the given insolvent Storage Provider. Once the total accrued file size reaches the total data size stored by the insolvent Storage Provider, the Storage Provider is deleted from the runtime.
+
+This process ensures that total redundancy is regained before the insolvent Storage Provider is removed from the network.
```

Comment:
Oh ok I see now, so the idea would be to give incentives to BSPs to provide the now lost redundancy for the files of the insolvent BSP so the process of regaining that redundancy is quicker, and we also now when that process is done so we can safely delete the BSP. Is that correct?

Reply:
Yes exactly, but anyone can call the `add_redundancy` extrinsic for lost files and get rewarded for it. For files that already exist, it must ofcourse the owner of the file to add more redundancy.